core/arch/arm/kernel/ree_fs_ta.c

```diff
@@ -240,15 +279,65 @@ static TEE_Result ree_fs_ta_read(struct user_ta_store_handle *h, void *data,
 
 	if (handle->offs + len > handle->nw_ta_size)
 		return TEE_ERROR_BAD_PARAMETERS;
-	if (data) {
+
+	if (handle->shdr->img_type == SHDR_ENCRYPTED_TA) {
+		if (data) {
+			dst = data; /* Hash secure buffer */
+			res = tee_ta_decrypt_update(handle->enc_ctx, dst, src,
+						    len);
+			if (res != TEE_SUCCESS)
+				return TEE_ERROR_SECURITY;
+		} else {
+			size_t num_bytes = 0;
+			size_t b_size = MIN(1024U, len);
+			uint8_t *b = malloc(b_size);
+
+			if (!b)
+				return TEE_ERROR_OUT_OF_MEMORY;
+
+			dst = NULL;
+			while (num_bytes < len) {
+				size_t n = MIN(b_size, len - num_bytes);
+
+				res = tee_ta_decrypt_update(handle->enc_ctx, b,
+							    src + num_bytes, n);
+				if (res)
+					break;
+				num_bytes += n;
+
+				res = crypto_hash_update(handle->hash_ctx, b,
```

Comment:
Since we're using authenticated encryption, why are we hashing too?

Reply:
IMO, authenticated encryption of a TA isn't self sufficient to provide strong authentication and non-repudiation properties on behalf of Service Provider (who owns the TA) since it may be possible for a device to self-sign a TA using authenticated encryption. So I think its required to do RSA signature verification as well where only Service Provider owns the private key to sign a TA.

Also, I see authenticated encryption of a TA to provide additional integrity of encrypted TA blob.

Reply:
If stronger integrity is needed it should be done by selecting a stronger hash function.

We need the hash as that's part of the signature, but it should be enough for it to only cover `struct shdr || struct shdr_bootstrap_ta || struct shdr_encrypted_ta`. If you're covering the entire binary AES-GCM is pointless and AES-CTR could be used instead.

Reply:
> We need the hash as that's part of the signature, but it should be enough for it to only cover struct shdr || struct shdr_bootstrap_ta || struct shdr_encrypted_ta.

As mentioned earlier, we would like to cover TA binary as well as that's something which needs to be authenticated to Service Provider.

> If you're covering the entire binary AES-GCM is pointless and AES-CTR could be used instead.

Using plain encryption AES-CTR with hash based verification approach is something similar to `Mac-and-Encrypt` which doesn't provide integrity of Ciphertext which maybe vulnerable to some chosen-ciphertext attacks (CCAs). Have a look at this [paper](https://iacr.org/archive/crypto2001/21390309.pdf). And the most recommended approach in `Encrypt-then-MAC` which is provided via AES-GCM approach.

Reply:
> As mentioned earlier, we would like to cover TA binary as well as that's something which needs to be authenticated to Service Provider.

You mean that the Tag and IV isn't enough to guarantee the integrity?

> Using plain encryption AES-CTR with hash based verification approach is something similar to Mac-and-Encrypt which doesn't provide integrity of Ciphertext which maybe vulnerable to some chosen-ciphertext attacks (CCAs). Have a look at this paper. And the most recommended approach in Encrypt-then-MAC which is provided via AES-GCM approach.

[I have to admit that I don't understand how this `ENC*` applies here, we're not doing the `*` transformation before encryption here.]

Change the order to Encrypt-then-Mac which seems like a more natural order any way. AES-CTR doesn't have to be used either, I only mentioned it since it's part of AES-GCM.

My main concern is that we should do better than something than just gets the job done, it should be efficient and elegant too.

Reply:
> > As mentioned earlier, we would like to cover TA binary as well as that's something which needs to be authenticated to Service Provider.
> 
> You mean that the Tag and IV isn't enough to guarantee the integrity?

Signing only Tag and IV aren't enough as it will be more like signing the encrypted blob (as Tag represents a unique value corresponding to encrypted blob) which doesn't provides non-repudiation property because only the illegible ciphertext, not the plaintext, would be non-repudiable. For more details, refer to this [link](http://world.std.com/~dtd/sign_encrypt/sign_encrypt7.html).

> My main concern is that we should do better than something than just gets the job done, it should be efficient and elegant too.

I agree with your concern and this is something that came initially to my mind too while researching for appropriate method to enable encrypted TAs. Let me try to elaborate more on the required security properties as follows that lead me to take approach implemented in this PR:

1. Authentication and non-repudiation of TA (in plain binary form) to Service Provider (SP).
2. Integrity of encrypted TA blob.

To satisfy (1), we need SP to sign TA (in plain binary form) that why hashing the complete plain TA and verification of signature using SP public key with hash is required.

To satisfy (2), we need to use MAC protection of encrypted blob provided by AES-GCM algo.

If we can still come up with more efficient approach to satisfy above security properties than I am open to using that approach.

Reply:
OK, makes sense. However I think that if the symmetric AES-GCM key was hashed too we wouldn't need to hash the plaintext. On the other hand, this seems to be a minefield so if there's a documented secure pattern we're better off using that rather than inventing our own.

I think this deserves a comment in the description of the header how the protection works and why it was selected.

Reply:
Sure, will add a comment in the description of the header.